### DISCUSSION TITLE
Forward and backward button in tabline

### DISCUSSION THREAD

**Discussion by the-c0d3r** [NvChad/NvChad#1751] [Show and tell]
![Screenshot from 2023-01-24 07-45-07](https://user-images.githubusercontent.com/4526565/214177340-840534e3-a0fb-4028-9771-e84be4c8ebf0.png)

I got feedback from a few newbie users that my neovim configuration is lacking a way to go back and forth, if they are jumping to different files through "go to reference" and such. And since they are newbie, they wanted to have a button that they can click, just like in IDEs, instead of having to press <kbd>Ctrl</kbd>+<kbd>i</kbd> or <kbd>Ctrl</kbd>+<kbd>o</kbd>.

I spent some time and managed to get those buttons in the tabline. I just wanted to share it for those who wanted the same thing as well.

It seems like `normal <C-i>` have a problem, whereby it seem to be inserted as tab. Luckily I found the solution in [this stackoverflow answer](https://vi.stackexchange.com/a/7553)

```lua
-- ~/.config/nvim/custom/buttons.lua
local tabufline_modules = require "nvchad_ui.tabufline.modules"

vim.cmd "function! JumpBackward(a,b,c,d) \n lua vim.cmd(vim.api.nvim_replace_termcodes('normal <C-o>',true,true,true)) \n endfunction"
vim.cmd "function! JumpForward(a,b,c,d) \n lua vim.cmd(vim.api.nvim_replace_termcodes('normal 1<C-i>',true,true,true)) \n endfunction"

return {
    buttons = function()
        local jmp_fwd_btn = "%@JumpForward@%" .. " › " .. "%X"
        local jmp_bwd_btn = "%@JumpBackward@%" .. " ‹ " .. "%X"
        return jmp_bwd_btn .. jmp_fwd_btn .. tabufline_modules.buttons()
    end,
}
```
I have no idea what kind of syntax is the `%@JumpForward@%` is using. I did some trial and error through the Nvchad UI repo's code. 

Then the chadrc.lua file

```lua
M.plugins = {
    override = {
        ["NvChad/ui"] = {
            statusline = {
                separator_style = "block", -- default/round/block/arrow
            },
            tabufline = {
                enabled = true,
                lazyload = false,
                overriden_modules = function()
                    return require "custom.buttons"
                end,
            },
        },
...
```

Lastly, feel free to suggest any improvement on the above configuration. It would be great if there are detailed instructions in the documentation to help users create new buttons and new functionality on the tabline.

**Comment by siduck** [comment]
@the-c0d3r i think tab & Shift + tab is much faster

**Reply by the-c0d3r** [reply]
The back and forth is to actually jump to back and forth on jump list position, rather than tab navigation.
Does `tab` and `shift+tab` follows the jump list or just to different tabs?
What I was looking for is a way to jump to next/previous locations even in the same file.

**Reply by siduck** [reply]
> The back and forth is to actually jump to back and forth on jump list position, rather than tab navigation. Does `tab` and `shift+tab` follows the jump list or just to different tabs? What I was looking for is a way to jump to next/previous locations even in the same file.

in nvchad tab & shift+tab will go to previous//next tab so it overrides the default behaviours. You could have overrided tabufline mappings to achieve your behavior

**Reply by the-c0d3r** [reply]
I see, but I wanted to have ability to do both, tab navigation (tab, shift tab), as well as jumping to locations in jump list (ctrl i, ctrl o). Just like how vim has two different jump functions.

Can I ask you how I can go about configuring colours and such for the buttons? I see this in [this file](https://github.com/NvChad/ui/blob/main/lua/nvchad_ui/tabufline/modules.lua#L220) but I don't know what kind of syntax is it, I only know the first portion `TbCloseAllBufs` is a vim function declared above, but I cannot find reference to `TbLineCloseAllBufsBtn` anywhere. 
```
  local CloseAllBufsBtn = "%@TbCloseAllBufs@%#TbLineCloseAllBufsBtn#" .. "  " .. "%X"
```

But if I just put any value inside `#`, like `#test#`, it still give me a gray colour background. How do I see a list of all the colours and apply them to the buttons?

**Reply by siduck** [reply]
@the-c0d3r  `"%#TblineTabNewBtn#"` , add this before the icon string of your buttons,    replace it with your own highlight group so the syntax is 

`"%#HighlightGroupName#mytext"`

**Reply by the-c0d3r** [reply]
Thanks for the info. I will try it out.

**Reply by siduck** [reply]
> Thanks for the info. I will try it out.

and define your hlgrou in chadrc

```lua
M.ui  = {

   hl_add = {  
   
       SomeBtns = {
              bg = "blue" , fg =  "black"    
       }
            
   }
            
}
